crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space/rpc_multi_logic.rs

```diff
@@ -219,15 +242,20 @@ impl Outer {
         let remote_request_grace_ms = self.remote_request_grace_ms;
         let agg = self.agg.clone();
         let grace_rs = self.grace_rs.clone();
+        let kill = self.kill.clone();
         Arc::new(move || {
             let permit = Share::new(grace_rs.acquire());
             let permit2 = permit.clone();
+            let kill = kill.clone();
 
             // the permit will exist for max grace period
             agg.push(
                 async move {
-                    tokio::time::sleep(std::time::Duration::from_millis(remote_request_grace_ms))
-                        .await;
+                    let f = tokio::time::sleep(std::time::Duration::from_millis(
+                        remote_request_grace_ms,
+                    ))
+                    .boxed();
+                    let _ = futures::future::select(f, kill.wait().boxed()).await;
```

Comment:
With rust futures often not being cancel safe, I feel like every select we use might need a disclaimer:

```suggestion
                    // This select is safe because we don't care if the timeout
                    // or the kill notifier get cancelled.
                    let _ = futures::future::select(f, kill.wait().boxed()).await;
```